Fix quantum keymapextra italian

sigul opened the pull request:
I reviewed and fixed a few definitions and added a few useful that were not present.

## Description

- change the duplicated definition of IT_PIPE, in favor of IT_DEGR based on the legend, and change the corresponding keycode according to that
- change IT_LESS form KC_NUBS to KC_GRAVE, as it was wrong
- add IT_TILDE, that was missing and it's useful to have especially on a Mac.
- add IT_ACUT definition, that was missing
- add missing legend for accented vowels
- rewrite left and right brackets (LBRC and RBRC) to shift + curly brackets (LSFT(LCBR), LSFT(RCBR)), based on the fact that to get them on the Italian keyboard you have to press alt + shift + brackets.
- add plus/minus key, not strictly necessary, but it was a bit difficult to find it, and thought it could be useful to someone to have it available
- reorder the file for readability

This was tested with my Planck keymap you can find here https://github.com/sigul/qmk_firmware/tree/sigul-planck/keyboards/planck/keymaps/sigul

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [x] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR
I discovered that using IT_ACUT on my keymap provoked an error because of the missing definition.
* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

fauxpark commented:
I think the "wrong" keys are actually correct, you just don't have the right keyboard layout set in macOS.

Here's the on screen keyboard in Windows, with the Italian layout:
![win](https://user-images.githubusercontent.com/4781841/65367235-a9d16b80-dc71-11e9-95a5-649fce01bb25.png)
And in macOS, after pressing the "Change keyboard type" button to convince it that my ANSI board is ISO:
<img width="390" alt="Screen Shot 2019-09-21 at 1 07 02 pm" src="https://user-images.githubusercontent.com/4781841/65367254-ddac9100-dc71-11e9-9be6-c1c309759163.png">

sigul commented:
@fauxpark i didn’t consider macOS would have act differently. Do you think we need an alternative file or can I simply create a section for Macos specific keys? I will check my keybaord on iPad for more information, not sure a test in a VM can help but will try. I tested with a Planck so you could be right. But I won’t say a keybaord will act differently based on the os, will it?

When it comes to the wrong key, I marked wrong only IT_LESS that was referring to KC_NUBS (non US backslash) that in the image you posted is on the upper left key, where usually you have ESC or GRAVE in US keyboards. I based my decision on the output I got from my keyboard when the MAC uses the Italian. 

I have an Anne Pro I can try to flash with qmk to test. Not sure it’s a good idea but will try.

fauxpark commented:
The existing convention is to have a separate file specifically for the macOS layout, yes. See the French and German ones.

> When it comes to the wrong key, I marked wrong only IT_LESS that was referring to KC_NUBS (non US backslash) that in the image you posted is on the upper left key, where usually you have ESC or GRAVE in US keyboards.

In the QMK configurator keyboard tester, when I press this key, it prints `KC_GRV`, and produces the backslash character, which is correct. The issue you have is that macOS thinks you're using an ANSI keyboard with the Italian layout, and changes the layout around a bit due to the missing NUBS key.

When you plug in a new keyboard macOS will prompt you to press the keys next to the Shift keys, to deduce what physical layout it has. Probably, if you cancel out of it, the default is ANSI. It may be kind of confusing when you have a board that is neither ANSI nor ISO like the Planck, but if you just mash the same button a couple of times, it will give up and ask you which one you want. Windows does not really have this problem (or feature?), as it blindly assumes your keyboard to be whatever physical layout that language typically uses.

As for the Anne Pro, I don't believe it is supported at the moment. It does use an STM32 chip (two of them, actually), but it's not one that I've seen before in the code.

sigul commented:
Ah I missed it, I will rename the file and add it as macos, while keeping the useful fix in this one. 

I will check the qmk configurator tester, didn’t think to check with it thanks. 

In the case of the Planck (any othto?) the layout should be set to ANSI, but in effect it makes more sense to set the layout based on the language of the device. 

In my case, my Mac never asked to press any key to detect the keyboard. I think it requires some more investigation on my side. Thanks @fauxpark to point me in the right direction

sigul commented:
@fauxpark I double checked the keyboard layout on my Mac with a Planck, a Contra, and a Anne Pro. As soon as I set Italian as a language, the Mac set up an ISO layout, but as soon as I start typing it switches to ANSI layout. If I run a test to recognize the keyboard, it fails with Contra and Planck because ortho misses the right shift. 

I am a bit confused now. 

I have this file that is used to have a Planck with an "ANSI Italian" layout (by logic, useful for anyone who wants to use an ANSI keyboard with a language that is ISO by default). I need this because otherwise my Planck won't work as I want, and will require me to define the keyboard layout each time I plug it in in an Apple device. That said, anyone who want to use an ANSI keyboard or ortho with Italian (European?) language on his device will need this additional file. At least, on an Apple device. Could you confirm? 

That said, considering that if you use an ANSI or ortho keyboard the Mac will automatically recognize it as ANSI, you will need an ANSI Italian additional file. This is why it took to me more than one year to make it, and lots of people keep trying! if you have a ISO layout you need the original keymap_italian.h (with some fixes, at least the double PIPE definition and the missing IT_ACUT). 

Should I submit a PR that rename the original keymap_italian.h to keymap_italian_ISO.h and adds keymap_italian_ANSI_MAC.h (or Apple instead of Mac, since it should work the same on the iPad, I can check)? Or just add my italian.h as ANSI_italian?

fauxpark commented:
> As soon as I set Italian as a language, the Mac set up an ISO layout, but as soon as I start typing it switches to ANSI layout. If I run a test to recognize the keyboard, it fails with Contra and Planck because ortho misses the right shift.

Yes, you basically have to press a key repeatedly and it will say it couldn't detect the layout, and ask you to choose it manually, as I mentioned. Also of note is that the physical layout setting is per-device, and you can see it switch if you have a board set up as ANSI and one as ISO when you type on each. On Macbooks it is hardcoded and you can't change it; the "Change keyboard type" button doesn't appear until you start typing on the external keyboard.

> Should I submit a PR that rename the original keymap_italian.h to keymap_italian_ISO.h and adds keymap_italian_ANSI_MAC.h (or Apple instead of Mac, since it should work the same on the iPad, I can check)? Or just add my italian.h as ANSI_italian?

That would be a breaking change. The best solution I think is to make two new files `keymap_italian_osx_iso.h` and `_ansi.h`.

sigul commented:
After some work (see below) I realized there's no ansi or iso issue, but just a osx one in case. Not even sure of this. but I reverted the file I edited and created two osx version, one with the file I proposed as osx_ansi.h and another osx_iso.h that is a clone of the original with some modifications according with what below (at first glance, it ended up being the same). I am posting on reddit for discussion.

-------

so, from my last edit,  I did cp keymap_italian.h to keymap_italian_osx_ansi.h and reverted the keymap_italian.h to the commit before my edit. I assume my old file just needs a statement to alert people this will work on a Mac with an Ansi / ortho layout.

When it comes to the file I wanted to edit, I reverted it to the commit before my editing, and copied it renaming it keymap_italian_osx_iso.h. I am now a bit confused and will let editing this file to someone with Windows. 

Here's what I did to add _osx_iso.h (assuming my file _osx_ansi.h is ok for ANSI keyboard on the Apple devices)

I played with ANSI Anne Pro and Macbook ISO around with the QMK Configurator keymap tester. With both this rules is ok (not sure about Windows)

    #define IT_LESS KC_GRV // < and >

I renamed the IT_PIPE to IT_DEGR and changed the definition (this should work even on Windows)

    #define IT_PIPE LSFT(IT_AACC) // °

With my MacBook keyboard (ISO), the \ and | in Italian are printed with IntBackslash that I assume is the KC_NUBS so it would be in the _osx_iso.h (I don't have this key on the Anne Pro so cannot double check)

    #define IT_BKSL KC_NUBS // 

Additionally IT_ACUT was missing and found it out being this

    #define IT_ACUT ALT(KC_8) // ´

and the grave being given by a different combination

    #define IT_GRAV ALT(KC_9) // `

Additionally, the missing Tilde

    #define IT_TILDE ALT(KC_5) // ~

I then realized the change I did are the same. So I am considering having just a osx file. 

Still, I don't realize how a keyboard should output different key based on the os, but will study it.

fauxpark commented on quantum/keymap_extras/keymap_italian_osx_ansi.h:
```diff
@@ -0,0 +1,112 @@
+#define IT_ACUT LALT(KC_8)     // ´
+#define IT_GRAVE LALT(KC_9)    // `
+
+#define	IT_X_PLUS X_RBRACKET  // #
```
```suggestion
```

These can be removed; the proper way to support other language layouts for sendstring is with a `sendstring_<language>.h` file.

sigul replied:
this was present in the original file quantum/keymap_extra/keymap_italian.h I edit, so I didn't touch it nor I know what it does. So, fro me it's ok to remove it, if you says. should I commit from here?

fauxpark replied:
Since these are new files you're adding, there's no need to keep them. It doesn't appear to be used anywhere at all, and I'll be moving the one in the original file to a "deprecated" section when I clean up this directory anyway.

fauxpark commented on quantum/keymap_extras/keymap_italian_osx_iso.h:
```diff
@@ -0,0 +1,110 @@
+#define IT_ACUT ALT(KC_8)      // ´
+#define IT_GRAV ALT(KC_9)      // `
+
+#define	IT_X_PLUS	X_RBRACKET 	// #
```
```suggestion
```

sigul commented:
I realized there's absolutely no need to have two different file for ansi and iso and merged them into one keymap_italian_osx.h and removed the not necessary definition indicated by @fauxpark

sigul commented:
@fauxpark Hi, I think this is ok now. Since I submitted it with the intent of editing a file, and ended up creating a new one, should I close this PR and open a new one, or is it ok to merge this?

fauxpark commented:
It still has the `\` and `<` keys swapped. If you could fix those, it should be good to merge.

sigul commented:
@fauxpark Here's why

With the **Mac set to use Italian,** I went to QMK Configurator Keyboard Test and run those tests

1) With my MacBook ISO keyboard 
IntBackSlash sends \  
KC_GRV sends < 

2) With the Anne Pro (ANSI)
KC_GRV sends <
KC_NUBS aka IntBackslash is missing

3) With my Planck, compiled used the file I submitted
KC_GRV sends < and > when shifted
KC_NUBS sends \ and | (IntBackslash)

Additionally, it sounds reasonable that KC_NUBS sends IntBackslash. 

From my side, I am pretty sure I am right, because it works on my Mac and my iPad Pro. This is why I accepted to rename the file with the appendix osx. I mean, I spent one year trying to make a custom layout taking for good that keymap_italian.h was good, and never being able to have a keyboard working as programmed. Then I checked the file, and found out it has something wrong. Fixed it, and now it's working for me. If this doesn't work on Windows, I accept, but on my 2 Macs it works. Now, maybe it's something related to my settings (I have English for default language and Italian for keyboard), but I really don't know what else I can do, and since moreover it make sense that non US backslash is equivalent to International backslash, I think at least on this I am right. 

Honestly, I don't really know what to do. I think this modifications could help others trying to create a layout in Italian, but I am ok keeping the file in my user space if you still think I am wrong. I want to stress that I am ok and there's no polemic intent, I just really would like to contribute to the project, but I don't have a Windows computer to test the file I propose on it.

fauxpark commented:
The issue is, as I said above, macOS has two different variants for each language, depending on whether you've told it that the keyboard is ANSI or ISO: 
![](https://i.stack.imgur.com/qbP3y.png)

So you need two of these keymap files to support both of them, one with `KC_GRV` mapped to `<` and no `KC_NUBS` (`_ansi.h`), the other with both, but mapped to the same characters as the existing Windows file (`_iso.h`).

sigul commented:
Not sure I understood.

You say that for Italian we need:

- keymap_italian.h
#define IT_LESS KC_NUBS
#define IT_BKSL KC_GRAVE 
- keymap_italian_osx_ansi.h 
#defineI T_LESS KC_GRV
#define IT_BKSL KC_NUBS (or nothing?)
- keymap_italian_osx_iso.h
#define IT_LESS KC_NUBS
#define IT_BKSL KC_GRAVE 

then I pick the file to include in the keyboard keymap based on? will an ANSI keyboard required to include both italian.h and italian_osx.h to be able to work in both windows and mac? this is really not clear to me.

My point here is that the keyboard of my MacBook, ISO layout, in Italian, when tested on QMK Configurator use
KC_GRV for < and >
KC_NUBS for \ and |

so I took this as a rule to define the required keys for my layout

sigul commented:
@fauxpark I think I found the issue.

[here](https://beta.docs.qmk.fm/faq/faq_keymap) you have KC_GRV where Mac Italian Keyboard has KC_NUBS, and KC_NUBS where the Mac Italian Keyboard has < and >, but

when testing the Mac Italian Keyboard on QMK configurator, the key with \ and | is identified as KC_NUBS (while the image on QMK state it is KC_GRV) and the key with < and > is identified as KC_GRV (while QMK doc identifies it as KC_NUBS). This doesn't change if you use an ANSI or ISO keyboard, KC_NUBS always outputs \ and | on the Mac when the language is set to Italian, and KC_GRV always outputs < and >. 

So, in my opinion, it would be correct to set KC_GRV to IT_LESS and KC_NUBS to IT_BKSL with no difference between iso and ansi. IMHO, this has nothing to do with ISO or ANSI, and we just need a file for osx (or eventually, to fix the existing italian keymap, but I can't check on Windows). Could you run a test with an ISO / ANSI keybaord on a Mac/ PC with language set to Italian to confirm or not?

fauxpark commented:
This is what I see in the Configurator, when hitting the top left key (on my board, this is programmed to `KC_GRV`):
<img width="724" alt="Screen Shot 2019-09-28 at 9 24 27 am" src="https://user-images.githubusercontent.com/4781841/65807510-2d3f0f80-e1d2-11e9-9900-e36d4cdc4bca.png">
<img width="724" alt="Screen Shot 2019-09-28 at 9 25 11 am" src="https://user-images.githubusercontent.com/4781841/65807511-2d3f0f80-e1d2-11e9-824c-a46ab3199fe7.png">
I didn't reflash or anything, just switched the keyboard type in the Keyboard preferences between ANSI and ISO. My board is *physically* ANSI but there is no way for macOS to know this, thus you have to set it yourself.

The existing `keymap_italian.h` is correct for both Windows (at all times, it is always ISO) and macOS (with the correct keyboard type set up for the board), until you want the additional Alted characters macOS gives you.

sigul commented:
man, I am going crazy with this. 

the original keymap_italian.h states
#define IT_LESS KC_NUBS
#define IT_BKSL KC_GRAVE

as per your test this is ok for only for an ISO keybaord.

KC_GRV sends < and > if the keyboard is set to be ANSI and language Italian.

so we need an italian_ansi.h in which KC_GRV sends < and >

am I right?

fauxpark commented:
Correct, but we also need an ISO one with the same macOS-only keycodes, and they should be named `_osx_iso.h` and `_osx_ansi.h`.

sigul commented:
So Italian.h is ok (just need to fix IT_ACUT missing definition and IT_PIPE that is defined 2 times, on of this for a missing IT_DEGR) and work for ISO keyboards. 

You want italian_osx_iso.h that is a clone of italian.h but also has `IT_PLMN`, `IT_TILDE`, `IT_ACUT` and `IT_GRAVE`

And an italian_osx_ansi.h (my file) with IT_LESS defined with KC_GRV and backslash with KC_NUBS

Still don’t understand the why, but if it’s what we need I do it.

fauxpark commented:
Pretty much, but ANSI keyboards don't have `KC_NUBS` or `KC_NUHS` so you can omit them there.

sigul commented:
I don't think so. KC_BSLS produces ù with ANSI keyboard and language set to Italian on MAC os, so I had to us KC_NUBS to get \ and |

![Screen Shot 2019-09-29 at 16 50 35](https://user-images.githubusercontent.com/7622916/65834239-5f876480-e2d9-11e9-8323-d06d8e7ca177.png)

I added the 3 edited files. Let me know if now them works

fauxpark commented:
> KC_BSLS produces ù with ANSI keyboard and language set to Italian on MAC os, so I had to us KC_NUBS to get \ and |

That's correct, but where would you put that key in an ANSI layout? If you want backslash and pipe, you should change the keyboard type to ISO and use the ISO file.

sigul commented:
Wherever he wants :)

It does not compromise the use of the ansi.h, just adds an option you will want to use if you have an ANSI board and want to set the language to Italian on your device, writing your keymap to simply add accented vowels on an ANSI layout. 

This was my starting point for my Planck, which could be used both as ANSI or ISO, but as soon as I plug it in in my Mac it is recognized as an ANSI board, modifying the layout for the Italian in the settings / input source to the one that generated this long thread (LESS in the upper left instead of backslash). So, basically, why should I tell to my Mac or iPad that my ortho board is an ISO if it sees it as ANSI and it's not one nor another? I assume that every newbie like me when starting writing code for his board won't go in the prefs to tell his Mac he's got an ISO board if he has an ortho.

So, I suggest to keep IT_BSLS mapped to KC_NUBS even in the keymap_italian_osx_ansi.h

fauxpark commented:
I don't know how exactly this should be dealt with when making a visual representation of the layout for the docs, as there's going to be an extra keycode defined that isn't present in the image. It could be slightly confusing. I think, since it "doesn't exist" on ANSI, it does not deserve a language-specific keycode. You can use `KC_NUBS` directly -- after all, it *is* called `Non-US Backslash`, and although that name is not always accurate for other languages it happens to output a backslash for Italian. But for now, it's fine.

fauxpark commented:
Thanks, and sorry for all the trouble!

sigul commented:
> Thanks, and sorry for all the trouble!

that was awesome. Thanks for helping me improving the code on this point. I really appreciated it.
